### DISCUSSION TITLE
error while loading giv image on linux

### DISCUSSION THREAD

**Discussion by ShiranGlasser** [dov/giv#68] [Q&A]
Hi @dov, since you helped me with the installation, I didn't try to use it till now because i was switching computers..
 I've tried to load an giv image on ubuntu and got this error:
`giv: error while loading shared libraries: libgiv-widget.so: cannot open shared object file: No such file or directory`
could you help me find the file?
the giv is istalled in the path: /usr/local/bin/giv

Thank you, Shiran

**Comment by dov** [comment]
Hi Shiran,

Short answer:
```
export LD_LIBRARY_PATH=/usr/local/lib64:/usr/local/lib
```

Longer answer: In contrast to Windows which uses the same environment variable `PATH` for executables and dynamic link (also called shared) libraries, on Linux, they reside in the two different variables, `PATH` and `LD_LIBRARY_PATH`. Further by linux convention, dynamic libraries are installed in .../lib or .../lib64 depending on the system. giv uses a couple of dynamic libraries that are installed in /usr/local/lib64 . And in order for the OS to find these, you must define LD_LIBRARY_PATH. (Or modify the ldconfig configuration, see `man ldconfig`, to modify it the library path system wide).

See also: https://www.geeksforgeeks.org/how-to-set-the-environmental-variable-ld_library_path-in-linux/

בהצלחה!

**Reply by ShiranGlasser** [reply]
@dov  Thank you very much, I'm not used to working on linux..
In my case the libgiv-widget.so was on /usr/local/lib/x86_64-linux-gnu
I used find to get the exact path and added it to LD_LIBRARY_PATH as you showed- 
it solved it, thank you!

**Comment by ShiranGlasser** [comment]
@dov  Hi again.
I now have the same issue when running the giv command while opening a giv image inside  `subprocess.Popen` on python, linux.
in the terminal it's working, but when executing the command with subprocess it doesn't recognize the libgiv-widget.so path again.
how do I add the path so the python subprocess will recognize it without adding script that sources the .bashrc file each time the code runs?

**Reply by dov** [reply]
Hi @ShiranGlasser . You just need to make sure that the environment variable is set before doing the `subprocess.Popen()`. Environment variables are inherited by a child process from its parent process. Here are a few ways that you can do to set it:

1. set the environment variable in the shell before running python, or if you are running python within an IDE, before running the ide.
2. Set the environment variable within python by `os.environ['LD_LIBRARY_PATH=']='/usr/local/...'`
3. Run subprocess open with the `shell=True` as follows: `subrocess.Popen('env LD_LIBRARY_PATH=/usr/local/...', shell=True)'
4. Another option if you have admin access is to create a file in '/etc/ld.so.conf.d/local.conf' and add the following line to it:
```
/usr/local/lib/x86_64-linux-gnu
```
then make sure to do `sudo ldconfig` or reboot your linux box.

Also double check that there is a match between the `LD_LIBRARY_PATH` and the location of the giv so files.
Good luck!
